miden-crypto/src/merkle/mmr/forest.rs

```diff
@@ -0,0 +1,458 @@
+use core::{
+    fmt::{Binary, Display},
+    ops::{BitAnd, BitOr, BitXor, BitXorAssign},
+};
+
+use super::InOrderIndex;
+use crate::Felt;
+
+/// A compact representation of trees in a forest. Used in the Merkle forest (MMR).
+///
+/// Each active bit of the stored number represents a disjoint tree with number of leaves
+/// equal to the bit position.
+///
+/// The forest value has the following interpretations:
+/// - its value is the number of leaves in the forest
+/// - the version number (MMR is append only so the number of leaves always increases)
+/// - bit count corresponds to the number of trees (trees) in the forest
+/// - each true bit position determines the depth of a tree in the forest
+///
+/// Examples:
+/// - `Forest(0)` is a forest with no trees.
+/// - `Forest(0b01)` is a forest with a single leaf/node (the smallest tree possible).
+/// - `Forest(0b10)` is a forest with a single binary tree with 2 leaves (3 nodes).
+/// - `Forest(0b11)` is a forest with two trees: one with 1 leaf (1 node), and one with 2 leaves (3
+///   nodes).
+/// - `Forest(0b1010)` is a forest with two trees: one with 8 leaves (15 nodes), one with 2 leaves
+///   (3 nodes).
+/// - `Forest(0b1000)` is a forest with one tree, which has 8 leaves (15 nodes).
+#[derive(Debug, Copy, Clone, Default, PartialEq, Eq, PartialOrd, Ord)]
+#[cfg_attr(feature = "serde", derive(serde::Deserialize, serde::Serialize))]
+pub struct Forest(usize);
+
+impl Forest {
+    /// Creates an empty forest (no trees).
+    pub const fn empty() -> Self {
+        Self(0)
+    }
+
+    /// Creates a forest with `num_leaves` leaves.
+    pub const fn new(num_leaves: usize) -> Self {
+        Self(num_leaves)
+    }
+
+    /// Creates a forest with a given height.
+    ///
+    /// This is equivalent to `Forest::new(1 << height)`.
+    ///
+    /// # Panics
+    ///
+    /// This will panic if `height` is greater than `usize::BITS - 1`.
+    pub const fn with_height(height: usize) -> Self {
+        assert!(height < usize::BITS as usize);
+        Self::new(1 << height)
+    }
+
+    /// Returns true if there are no trees in the forest.
+    pub fn is_empty(self) -> bool {
+        self.0 == 0
+    }
+
+    /// Returns a forest with a capacity for exactly one more leaf.
+    ///
+    /// Some smaller trees might be merged together.
+    pub fn append_leaf(&mut self) {
```

Comment:
I think the comment here needs to be updated as we don't really return anything from this function.